### PR TITLE
Fix silent loss of per-iteration logging in `GepaPromptOptimizer` when `valset` is passed via `gepa_kwargs`

### LINKED ISSUE #24461: [BUG] Silent loss of per-iteration logging in `GepaPromptOptimizer` when `valset` is passed via `gepa_kwargs`
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.14.1.dev0 (source install, master @ ad9efa076)
- Tracking server: no tracking server

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 24.04.3
- **Python version**: 3.10.20

### Describe the problem

`mlflow/mlflow/genai/optimize/optimizers/gepa_optimizer.py:196` distinguishes full-valset evaluations from reflective-mutation minibatch evaluations by batch size. But if len(valset) != len(trainset), valset eval never will be logged.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
# Repro: per-iteration logging is silently lost in GepaPromptOptimizer when a
# valset is passed via gepa_kwargs.
#
# OPENROUTER_API_KEY must be set (used only by GEPA's reflection model via
# litellm; predict_fn and the scorer are deterministic). Set
# OPENROUTER_API_BASE to override the default https://openrouter.ai/api/v1.

import os
import tempfile

import mlflow
from mlflow.genai.optimize.optimizers import GepaPromptOptimizer
from mlflow.genai.scorers import scorer

assert "OPENROUTER_API_KEY" in os.environ, "needed for the GEPA reflection model only"

# Hermetic tracking store so the repro is independent of any MLFLOW_TRACKING_URI
# set in the environment (the filesystem backend is in maintenance mode, so use sqlite).
mlflow.set_tracking_uri(f"sqlite:///{tempfile.mkdtemp(prefix='gepa-valset-repro-')}/mlflow.db")
mlflow.set_experiment("gepa-valset-repro")

prompt = mlflow.genai.register_prompt(
    name="qa",
    template="Answer the following question: {{question}}",
)

ANSWERS = {
    "What is the capital of France?": "Paris",
    "What is the capital of Germany?": "Berlin",
    "What is the capital of Italy?": "Rome",
    "What is the capital of Spain?": "Madrid",
    "What is the capital of Japan?": "Tokyo",
    "What is the capital of Canada?": "Ottawa",
    "What is the capital of Egypt?": "Cairo",
}


def predict_fn(question: str) -> str:
    prompt.format(question=question)
    return ANSWERS.get(question, "unknown")


@scorer(name="exact_match")
def exact_match(outputs, expectations):
    return 1.0 if outputs == expectations else 0.0


def record(question):
    return {"inputs": {"question": question}, "outputs": ANSWERS[question]}


train_data = [
    record("What is the capital of France?"),
    record("What is the capital of Germany?"),
    record("What is the capital of Italy?"),
    record("What is the capital of Spain?"),
]
# Deliberately a different size than train_data (3 != 4) — that is the trigger.
valset = [
    record("What is the capital of Japan?"),
    record("What is the capital of Canada?"),
    record("What is the capital of Egypt?"),
]


def run_case(gepa_kwargs):
    mlflow.genai.optimize_prompts(
        predict_fn=predict_fn,
        train_data=train_data,
        prompt_uris=[prompt.uri],
        optimizer=GepaPromptOptimizer(
            reflection_model="openrouter:/deepseek/deepseek-chat-v3-0324",
            max_metric_calls=10,
            gepa_kwargs=gepa_kwargs,
        ),
        scorers=[exact_match],
    )
    run_id = mlflow.last_active_run().info.run_id
    client = mlflow.MlflowClient()
    candidate_artifacts = [a.path for a in client.list_artifacts(run_id, "prompt_candidates")]
    metrics = client.get_run(run_id).data.metrics
    return candidate_artifacts, sorted(m for m in metrics if m.startswith("eval_score"))


artifacts, metrics = run_case(gepa_kwargs=None)
print(f"without valset: artifacts={artifacts} metrics={metrics}")

artifacts, metrics = run_case(gepa_kwargs={"valset": valset})
print(f"with valset:    artifacts={artifacts} metrics={metrics}")

# Expected: both runs log prompt_candidates/iteration_N/* artifacts and
# eval_score / eval_score.exact_match metrics (the valset run evaluated on
# the valset instead of the trainset).
#
# Actual:
#   without valset: artifacts=['prompt_candidates/iteration_0', ...] metrics=['eval_score', 'eval_score.exact_match']
#   with valset:    artifacts=[] metrics=[]
# No warning or error is raised — the per-iteration tracking is silently lost,
# because MlflowGEPAAdapter detects "full validation pass" by comparing the
# batch size to len(train_data), which never matches len(valset).
```

### Stack trace

```
REPLACE_ME
```

### Other info / logs

```
OPENROUTER_API_KEY=sk-or-...  uv run --frozen --with 'gepa>=0.0.26' --with 'litellm' python specs/gepa-va
lset-repro.py
Installed 58 packages in 49ms
2026/07/16 07:04:09 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/07/16 07:04:09 INFO mlflow.store.db.utils: Updating database tables
2026/07/16 07:04:14 INFO mlflow.tracking.fluent: Experiment with name 'gepa-valset-repro' does not exist. Creating a new experiment.
2026/07/16 07:04:14 INFO mlflow.genai.utils.data_validation: Testing model prediction with the first sample in the dataset. To disable this check, set the MLFLOW_GENAI_EVAL_SKIP_TRACE_VALIDATION environment variable to True.
/home/shpilevo/projects/mlflow/mlflow/data/dataset_source_registry.py:148: UserWarning: The specified dataset source can be interpreted in multiple ways: LocalArtifactDatasetSource, LocalArtifactDatasetSource. MLflow will assume that this is a LocalArtifactDatasetSource source.
  return _dataset_source_registry.resolve(
2026/07/16 07:04:18 INFO mlflow.models.evaluation.utils.trace: Auto tracing is temporarily enabled during the model evaluation for computing some metrics and debugging. To disable tracing, call `mlflow.autolog(disable=True)`.
Iteration 0: Base program full valset score: 0.0 over 4 / 4 examples
Iteration 1: Selected program 0 score: 0.0
Iteration 1: Proposed new text for qa: Answer the following question about the capital of a country. The question will be in the format: "What is the capital of [Country]?" Your response should be the name of the capital city only, without any additional text or punctuation. 

For example:
- Input: "What is the capital of Italy?" 
  Output: "Rome"
- Input: "What is the capital of Germany?" 
  Output: "Berlin"
- Input: "What is the capital of France?" 
  Output: "Paris"

Ensure your answer is accurate and concise. Do not provide explanations, just the name of the capital city.
Iteration 1: New subsample score 0.0 is not better than old score 0.0, skipping
without valset: artifacts=['prompt_candidates/iteration_0'] metrics=['eval_score', 'eval_score.exact_match']
2026/07/16 07:04:37 INFO mlflow.genai.utils.data_validation: Testing model prediction with the first sample in the dataset. To disable this check, set the MLFLOW_GENAI_EVAL_SKIP_TRACE_VALIDATION environment variable to True.
/home/shpilevo/projects/mlflow/mlflow/data/dataset_source_registry.py:148: UserWarning: The specified dataset source can be interpreted in multiple ways: LocalArtifactDatasetSource, LocalArtifactDatasetSource. MLflow will assume that this is a LocalArtifactDatasetSource source.
  return _dataset_source_registry.resolve(
Iteration 0: Base program full valset score: 0.0 over 3 / 3 examples
Iteration 1: Selected program 0 score: 0.0
Iteration 1: Proposed new text for qa: Answer the following question about the capital of a country. The response should be a single word or short phrase representing the capital city, without additional explanation or elaboration. Ensure the answer is factually accurate and up-to-date. 

Example inputs and outputs:
- Input: "What is the capital of Italy?" Output: "Rome"
- Input: "What is the capital of Germany?" Output: "Berlin"
- Input: "What is the capital of France?" Output: "Paris"

If the question is not about a country's capital or is unclear, respond with "I can only answer questions about the capitals of countries."
Iteration 1: New subsample score 0.0 is not better than old score 0.0, skipping
Iteration 2: Selected program 0 score: 0.0
Iteration 2: Proposed new text for qa: Answer the following capital city question with just the name of the city. 
The input will always be in the format "What is the capital of [Country]?" 
where [Country] is the name of a sovereign state. 
Respond only with the correct capital city's name - do not include any additional words, punctuation or formatting. 
Example inputs and correct outputs:
- Input: "What is the capital of Spain?" → Output: "Madrid"
- Input: "What is the capital of France?" → Output: "Paris"
- Input: "What is the capital of Japan?" → Output: "Tokyo"
Iteration 2: New subsample score 0.0 is not better than old score 0.0, skipping
with valset:    artifacts=[] metrics=[]
```

### What component(s) does this bug affect?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by PreethamSanji**
### Related Issues/PRs

Closes #24461

### What changes are proposed in this pull request?

**Root cause:** `MlflowGEPAAdapter.evaluate()` detects a "full validation pass" — the only case
where per-iteration artifacts and metrics are logged — by comparing the batch size against
`full_dataset_size`, which was hardcoded to `len(train_data)`. However, GEPA only validates on the
trainset when no `valset` is provided; when a `valset` is passed through `gepa_kwargs` (a documented
escape hatch, and `valset` is not among the reserved params overridden by MLflow), full validation
batches have `len(valset)` instead (see gepa's `api.py`:
`val_loader = ensure_loader(valset) if valset is not None else train_loader`). Whenever
`len(valset) != len(train_data)`, the size check never matches, `_log_validation_candidate` is never
called, and all `prompt_candidates/iteration_N/*` artifacts and `eval_score` metrics are silently
dropped. Optimization itself succeeds, so users get no warning.

**Fix:** Derive `full_dataset_size` from the effective validation set, mirroring GEPA's own
fallback: `len(valset)` if a `valset` is present in `gepa_kwargs`, otherwise `len(train_data)`. The
adapter docstring is updated accordingly.

**Backwards compatible:** when no `valset` is passed, `full_dataset_size` resolves to
`len(train_data)` exactly as before; the default path is behavior-identical. Not a significant
change per CONTRIBUTING.md: no REST API changes, no new user-facing APIs, no new dependencies.

**Known residual limitation (pre-existing, out of scope):** the size heuristic can still misfire if
the effective valset size coincides with GEPA's reflection minibatch size (default 3). A more robust
detection would require a signal from GEPA other than batch length; happy to file a follow-up issue
if maintainers think it's worth tracking.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

New test `test_gepa_optimizer_logs_prompt_candidates_with_custom_valset` (parametrized with valset
both smaller and larger than the trainset) verifies that: (1) full validation passes over a custom
valset log the iteration table, artifacts, and step-wise metrics; (2) minibatch evaluations still
log nothing; (3) a trainset-sized batch is no longer misclassified as a full validation pass when a
custom valset is present. The existing `test_gepa_optimizer_logs_prompt_candidates` covers the
unchanged default (no-valset) path.

Manual test: a fully offline repro (mocked `gepa`, no LLM calls) that simulates GEPA's evaluation
contract against the adapter and counts logging calls.

Before this PR:
`no valset : log_metrics=1 log_table=1 log_artifact=2
with valset: log_metrics=0 log_table=0 log_artifact=0  <-- per-iteration logging LOST`

After this PR:
`no valset : log_metrics=1 log_table=1 log_artifact=2
with valset: log_metrics=1 log_table=1 log_artifact=2`

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixed `GepaPromptOptimizer` silently skipping per-iteration artifact and metric logging
(`prompt_candidates/iteration_N/*`, `eval_score`) when a custom `valset` is passed via
`gepa_kwargs`.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

File: tests/genai/optimize/optimizers/test_gepa_optimizer.py [hunk 1]
```diff
@@ -659,3 +659,97 @@ def test_extract_eval_scores_per_scorer(val_aggregate_scores, val_aggregate_subs
...
+            mock_result.best_candidate = sample_target_prompts
+            mock_result.val_aggregate_scores = [0.8]
+            mock_result.val_aggregate_subscores = None
+            return mock_result
+
+        mock_gepa_module.optimize = mock_optimize_fn
+
+        with mlflow.start_run():
+            with (
+                patch("mlflow.genai.optimize.optimizers.gepa_optimizer.mlflow.log_artifact"),
```
[the remark this answers is not in the document]
Comment (PreethamSanji) [reply tests/genai/optimize/optimizers/test_gepa_optimizer.py]: Understood and updated!

File: tests/genai/optimize/optimizers/test_gepa_optimizer.py [hunk 2]
```diff
@@ -659,3 +659,97 @@ def test_extract_eval_scores_per_scorer(val_aggregate_scores, val_aggregate_subs
...
+                # when a custom valset is provided, so it should NOT be logged
+                captured_adapter.evaluate(
+                    sample_train_data, {"system_prompt": "Test"}, capture_traces=False
+                )
+                assert logged_tables == []
+                assert logged_metrics == []
+
+                # Full validation pass over the custom valset SHOULD be logged
+                candidate = {"system_prompt": "Optimized prompt", "instruction": "New instruction"}
+                captured_adapter.evaluate(valset, candidate, capture_traces=False)
```
[the remark this answers is not in the document]
Comment (PreethamSanji) [reply tests/genai/optimize/optimizers/test_gepa_optimizer.py:750]: Updated as recommended

**Comment by PreethamSanji** [comment]
Hey! Friendly comment for check

**Comment by PreethamSanji** [comment]
@B-Step62 Requesting for review!

**Review by B-Step62 — APPROVED** [review]
LGTM, thank you for the fix!
